File: api/gitlab/gitlab.go
package gitlab

import (
	"bytes"
	"encoding/json"
	"errors"
	"fmt"
	"io/ioutil"
	"net/http"
	"net/url"
)

// Config holds the configuration needed to connect to the Gitlab API
type Config struct {
	URL         string
	AccessToken string
	ProjectID   string
}

// Issue represents the format of an Issue returned when getting an issue from the Gitlab API
type Issue struct {
	ID          int      `json:"id"`
	State       string   `json:"state"`
	Title       string   `json:"title"`
	Description string   `json:"description"`
	CreatedAt   string   `json:"created_at"`
	Weight      string   `json:"weight"`
	WebURL      string   `json:"web_url"`
	Labels      []string `json:"labels"`
}

// CreateIssueBody represents the request body used for creating an issue using the GitLab API
type CreateIssueBody struct {
	Title       string `json:"title"`
	Description string `json:"description"`
	Weight      string `json:"weight"`
	Labels      string `json:"labels"`
}

// CreateIssue creates an issue using the GitLab API and returns the created issue
func CreateIssue(config Config, issue CreateIssueBody) (Issue, error) {
	parsedURL, err := url.Parse(config.URL)
	if err != nil {
		return Issue{}, err
	}

	parsedURL.Path = "/api/v4/projects/" + config.ProjectID + "/issues"

	jsonBody, err := json.Marshal(issue)
	if err != nil {
		return Issue{}, err
	}

	req, err := http.NewRequest("POST", parsedURL.String(), bytes.NewBuffer(jsonBody))
	if err != nil {
		return Issue{}, err
	}

	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("PRIVATE-TOKEN", config.AccessToken)

	client := &http.Client{}

	resp, err := client.Do(req)
	if err != nil {
		return Issue{}, err
	}

	if resp.StatusCode != http.StatusCreated {
		buf, _ := ioutil.ReadAll(resp.Body)
		errorMessage := fmt.Sprintf("Did not receive a StatusCreated code from the Gitlab API, but received %d, message: %s", resp.StatusCode, buf)
		return Issue{}, errors.New(errorMessage)
	}

	getIssue := Issue{}
	err = json.NewDecoder(resp.Body).Decode(&getIssue)
	if err != nil {
		return Issue{}, err
	}

	return getIssue, nil
}

File: api/core/gitlab.py
import requests
from django.core.exceptions import ImproperlyConfigured
from django.conf import settings

REQUEST_TIMEOUT_SECONDS = 10


def create_issue(title, description, labels):
    if not settings.GITLAB.get('ACCESS_TOKEN'):
        raise ImproperlyConfigured('Gitlab access token is missing.')

    if not settings.GITLAB.get('PROJECT_ID'):
        raise ImproperlyConfigured('Gitlab project id is missing.')

    if not settings.GITLAB.get('URL'):
        raise ImproperlyConfigured('Gitlab URL is missing.')

    headers = {
        'Content-Type': 'application/json',
        'PRIVATE-TOKEN': settings.GITLAB['ACCESS_TOKEN'],
    }

    body = {
        'title': title,
        'description': description,
        'labels': labels
    }

    url = f'{settings.GITLAB["URL"]}/api/v4/projects/{settings.GITLAB["PROJECT_ID"]}/issues'

    response = requests.post(url, json=body, headers=headers, timeout=REQUEST_TIMEOUT_SECONDS)
    response.raise_for_status()

    return response.json()
